KIP-81: KAFKA-4133: Bound memory usage of the Consumer

mimaison opened the pull request:
Introduce new Consumer config 'buffer.memory'
Allocate MemoryPool based on 'buffer.memory'
Allocate NetworkReceive through the MemoryPool
Communication between client and Coordinator node skips MemoryPool
Expose MemoryPool metrics in consumer-metrics group

https://cwiki.apache.org/confluence/display/KAFKA/KIP-81%3A+Bound+Fetch+memory+usage+in+the+consumer

Co-authored-by: Mickael Maison <mickael.maison@gmail.com>
Co-authored-by: Edoardo Comar <ecomar@uk.ibm.com>

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

mimaison committed:
KIP-81: KAFKA-4133: Bound memory usage of the Consumer

Introduce new Consumer config 'buffer.memory'
Allocate MemoryPool based on 'buffer.memory'
Allocate NetworkReceive through the MemoryPool
Communication between client and Coordinator node skips MemoryPool
Expose MemoryPool metrics in consumer-metrics group

https://cwiki.apache.org/confluence/display/KAFKA/KIP-81%3A+Bound+Fetch+memory+usage+in+the+consumer

Co-authored-by: Mickael Maison <mickael.maison@gmail.com>
Co-authored-by: Edoardo Comar <ecomar@uk.ibm.com>

mimaison commented:
@hachikuji @rajinisivaram Can you have a look ?

Also when you decide to merge, can you ensure the `Co-authored-by` blocks in the commit messages are kept.

mimaison commented:
@hachikuji @rajinisivaram gentle reminder =)

mimaison commented:
@rajinisivaram I see this was moved out of 2.0. Any particular reason? We've had this ready to be reviewed for a month now

rajinisivaram commented:
@mimaison I think @ijuma had moved out some of the KIPs which didn't have any one available to review yet. @ijuma  Was this one of the KIPs you moved out of the release?

mimaison commented:
Due to the amount of files this is changing, I have to keep rebasing regularly.

@hachikuji @rajinisivaram Can you take a look?

hachikuji commented:
@mimaison Thanks for the updates and apologies for the delay. I will review today if possible, but early next week is probably more likely.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -748,13 +747,19 @@ private void maybeThrottle(AbstractResponse response, short apiVersion, String n
+            try {
+                receive.close();
+            } catch (IOException e) {
+                log.info("Failed closing NetworkReceive from {}", source);
```
Is it intentional to swallow this exception? At a minimum, we should probably log the exception at DEBUG level.

mimaison replied:
Yes we should log the exception

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -257,7 +257,13 @@
+
+    /** <code>buffer.memory</code> */
+    public static final String BUFFER_MEMORY_CONFIG = "buffer.memory";
+    public static final String BUFFER_MEMORY_DOC = "The total bytes of memory the consumer can use to buffer records received from the server and waiting to be processed (decompressed and deserialized)." +
```
nit: "and waiting" -> "while waiting"?

mimaison replied:
Fixed

hachikuji commented on clients/src/main/java/org/apache/kafka/common/Node.java:
```diff
@@ -28,20 +28,30 @@
     private final String host;
     private final int port;
     private final String rack;
+    private final boolean privileged;
```
I wonder if we can find a more descriptive name. Can't think of a great one, but maybe something like `allowMemoryOveruse`? At a minimum, we should add some documentation describing the intent of this flag.

mimaison replied:
I've left it as `privileged` at least for now but added some javadoc.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -2230,4 +2238,37 @@ private void throwIfNoAssignorsConfigured() {
+                return memoryPool.availableMemory();
+            }
+        };
+        metrics.addMetric(metrics.metricName("memory-pool-free",
```
We should probably update the KIP with these metrics.

mimaison replied:
Good point! I've added a small section in the KIP listing all 4 new metrics

hachikuji commented on clients/src/main/java/org/apache/kafka/common/memory/SimpleMemoryPool.java:
```diff
@@ -39,7 +39,17 @@
 
-    public SimpleMemoryPool(long sizeInBytes, int maxSingleAllocationBytes, boolean strict, Sensor oomPeriodSensor) {
+    public static MemoryPool create(long sizeInBytes, int maxSingleAllocationBytes, Sensor oomPeriodSensor) {
+        // -1 means unbounded on the broker
```
Mention these in a method javadoc?

mimaison replied:
Added javadoc

hachikuji commented on clients/src/main/java/org/apache/kafka/common/network/NetworkReceive.java:
```diff
@@ -108,10 +94,16 @@ public long readFrom(ScatteringByteChannel channel) throws IOException {
-            if (buffer == null)
-                log.trace("Broker low on memory - could not allocate buffer of size {} for source {}", requestedBufferSize, source);
+        if (buffer == null && requestedBufferSize != -1) { //we know the size we want but haven't been able to allocate it yet
+            if (privileged) {
```
My memory is a little hazy on this KIP, but I thought we would allocate from the memory pool if possible, but fall back to direct allocation if that fails.

mimaison replied:
Yes you're right. In the initial implementationI had made this simplification, but I agree let's use the MemoryPool if there's space

hachikuji commented on clients/src/main/java/org/apache/kafka/common/network/Selector.java:
```diff
@@ -176,20 +178,12 @@ public Selector(int maxReceiveSize,
-            ChannelBuilder channelBuilder,
-            LogContext logContext) {
-        this(maxReceiveSize, connectionMaxIdleMs, metrics, time, metricGrpPrefix, metricTags, metricsPerConnection, false, channelBuilder, MemoryPool.NONE, logContext);
+    public Selector(long connectionMaxIdleMS, Metrics metrics, Time time, String metricGrpPrefix, ChannelBuilder channelBuilder, MemoryPool memoryPool, LogContext logContext) {
```
For long argument lists, we typically prefer to have each argument on a separate line, as in the original code. It's a bit subjective, but I think it improves readability.

mimaison replied:
Done

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -2230,4 +2238,37 @@ private void throwIfNoAssignorsConfigured() {
     }
 
+    private MemoryPool createMemoryPool(Metrics metrics, String metricGrpPrefix, long memoryPoolSize, int maxFetchBytes) {
+        if (memoryPoolSize <= 0L || memoryPoolSize <= maxFetchBytes) {
```
I think technically the broker is allowed to exceed the max fetch bytes to ensure that at least one message fits into the response. I'm wondering how we would handle that in this patch?

mimaison replied:
The memory pool we create is not strict so it could be used to allocate more than its specified size. The issue is with `maxSingleAllocationBytes`. It's currently set to `maxFetchBytes` so as you pointed it's not going to work. 
Possible solutions:
- Set `maxSingleAllocationBytes` as a factor (x2 ?) of `maxFetchBytes`
- Set `maxSingleAllocationBytes` to max Int

hachikuji replied:
Since the default is unbounded, it may be reasonable to raise an exception to the user if a fetch response exceeds a custom configured `buffer.memory`.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/network/NetworkReceive.java:
```diff
@@ -108,10 +94,16 @@ public long readFrom(ScatteringByteChannel channel) throws IOException {
+                // If the channel is privileged, we allocate directly on the heap
+                buffer = ByteBuffer.allocate(requestedBufferSize);
+            } else {
+                buffer = memoryPool.tryAllocate(requestedBufferSize);
```
Related to a previous question, but what happens when the requested buffer size exceeds the capacity of the pool?

hachikuji reviewed (commented):
@mimaison Thanks for the patch. I left a few comments. 

There is one basic thing that I think I'm missing at a high level. I was expecting to see some code to deallocate buffers in `Fetcher` after the consumer had finished parsing them, but I couldn't find anything. Can you clarify how deallocation is supposed to work?

The other thing I'm wondering about is testing. How have you been able to validate that the memory usage actually is bounded?

mimaison commented:
@hachikuji Deallocation happens in NetworkClient when we close the NetworkReceive (https://github.com/apache/kafka/pull/4934/files#diff-3c7816029b61d22ff4b01c6fc5f61392R759). Passed this point, the records are returned to the client and considered user-owned memory

mimaison commented:
@hachikuji Thanks for the review! Let's get this in for the next release.
We've addressed most of the comments. There's still a few open questions. I hope I've provided ideas for solving them.
At the moment, only tests have conflicts, I'll rebase when we'll start looking at them

hachikuji commented:
@mimaison I'm still confused about the deallocation. The fetched data is buffered inside the Fetcher in the `completedFetches` collection. I had assumed that our memory bound would cover this, but it sounds like that is not the case? I actually wonder how this can work if it does not cover this because the FetchResponse has a reference to the underlying buffer that was received.

mimaison commented:
You're right it's currently not the case. I agree it would be better to count data buffered in the Fetcher in the memory pool as it's still client-owned.

By setting the original ByteBuffer to null, it does not updates the content of the ByteBuffer backing the completedFetch (that was created as a slice of the original). Only once the completedFetch goes out of scope, the underlying buffer will be collected.

hachikuji commented:
> You're right it's currently not the case. I agree it would be better to count data buffered in the Fetcher in the memory pool as it's still client-owned.

Yeah, I think the impact is significantly lessened if we do not account for this.

> By setting the original ByteBuffer to null, it does not updates the content of the ByteBuffer backing the completedFetch (that was created as a slice of the original). Only once the completedFetch goes out of scope, the underlying buffer will be collected.

I think this is true given the implementation of `SimpleMemoryPool`, but it seems brittle to assume that a ByteBuffer (or a slice of one) remains usable after a call to `MemoryPool.deallocate`.

mimaison commented:
> Yeah, I think the impact is significantly lessened if we do not account for this.

Agreed! In that case, we need to pass the NetworkReceive reference up to `CompletedFetch` so Fetcher can close it when it returns the `CompletedFetch` to the user.

> I think this is true given the implementation of SimpleMemoryPool, but it seems brittle to assume that a ByteBuffer (or a slice of one) remains usable after a call to MemoryPool.deallocate. Unfortunately we can't only pass the Buffer as we need to know if it was allocated in the Pool or not.

Well `SimpleMemoryPool` does not really deallocates anything. It just maintains a counter of things it considers to be the owner of. When we call. In our case, when we call `release()`, we simply tell the Pool the user now owns this piece of memory and it can stop caring about it.

hachikuji commented:
> Well SimpleMemoryPool does not really deallocates anything. It just maintains a counter of things it considers to be the owner of. When we call. In our case, when we call release(), we simply tell the Pool the user now owns this piece of memory and it can stop caring about it.

You are right, but my point is that this behavior is specific to the implementation of `SimpleMemoryPool` and not a contract of the `MemoryPool` interface. We should avoid tying the code too tightly to this implementation. In any case, I think we agree that this patch should cover the buffered data in `Fetcher`?

mimaison commented:
> In any case, I think we agree that this patch should cover the buffered data in Fetcher?

Yes, see my updated comment above (https://github.com/apache/kafka/pull/4934#issuecomment-411343633). My answer was wrongly formatted in the <quote> block and you did not spot it.

> You are right, but my point is that this behavior is specific to the implementation of SimpleMemoryPool and not a contract of the MemoryPool interface. We should avoid tying the code too tightly to this implementation.

So can we even call `release()` once the data has been given to the user? I'm not too familiar with the `ConsumerRecords` pipeline so I'll need to check if at any point the data returnedto users is not backed anymore by the same underlying `ByteBuffer`

hachikuji commented:
@mimaison At the moment, we do not return the buffer to the user. This has been suggested in the past, but currently not even deserializers get access to the raw buffer. It should be safe to release the buffer after we have consumed the fetch response inside `Fetcher`.

mimaison commented:
I've had a little bit of time today to take another look. I realized that once returned to the `Fetcher`, `ClientResponse` is split into several `CompletedFetch` which makes it tricky to track as we'd want to only deallocate once all `CompletedFetch` have been processed.
So far I've only moved the deallocation after the `ClientResponse` has been split.

edoardocomar commented:
Hi @hachikuji we've updated the way responses are deallocated. Based on the schema, we know which ones contain fields backed by a ByteBuffer and need to be deallocated after a full data consumption.
The FetchResponse is a special case as its buffer gets sliced into many CompletedFetches.

We've added a test that uses the `GarbageCollectedMemoryPool` (which we plug into the factory) to check for failed deallocations and reads after deallocation.

We'll post later some data about the memory usage of a Consumer with and without a MemoryPool.
Feedback welcome :-)

mimaison commented:
Tests were performed using:
- Kafka cluster (trunk) with 6 brokers (default configuration)
- 1 topic with 6 partitions (1 partition on each broker) filled with 100k messages

Running a consumer without a memory pool:
- 300MB heap (`-Xmx300M -Xms300M`)
- `fetch.max.bytes=32428800`
- `max.partition.fetch.bytes=15428800`
It OOMs within 15 seconds

Running a consumer with a memory pool (30MB):
- 300MB heap (`-Xmx300M -Xms300M`)
- `fetch.max.bytes=32428800`
- `max.partition.fetch.bytes=15428800`
- `buffer.memory=32428801`
It's running without any issues using less than 200MB of memory.
<img width="875" alt="screen shot 2018-09-20 at 15 37 08" src="https://user-images.githubusercontent.com/903615/45822411-b3856380-bceb-11e8-9952-f351f635ca53.png">

Both Consumers have the exact same logic, a call to `poll()` in a loop.

In both cases, each brokers return around 15MB of data (150 messages). 
- Without the pool, each `poll()` returns messages from 3 or more partitions, quickly using up all memory.
- With the pool, the 1st 2 FetchResponses exhaust the pool, keeping the other responses queued in the network buffer. Calls to `poll()` only returns messages from 1 or 2 partitions and only once records have been returned, the other responses are read. So the pace of the Consumer is slowed down but it is able to use significantly less memory.

edoardocomar commented:
Here is a result of the impact of the new setting `buffer.memory` in this other scenario:

Kafka cluster with 6 brokers
2 topics with 6 partitions each, one per broker
both topics well preloaded with messages of size ~100k 

Demo app with 1 consumer subscribed to both topics running a loop like
```
while(true) {
consumer.poll(100L)
Thread.sleep(1000L)
}
```
and consumer config 
```
"fetch.max.bytes"=10000000
"max.partition.fetch.bytes"=5000000
```
on Java 1.8.0_181 on MacOS

1) running this single consumer app without the memory pool (i.e. no `buffer.memory` set)
required  -Xmx200M -Xms200M  to avoid an OOM error.

![image](https://user-images.githubusercontent.com/6661532/45823430-9d74a480-bce5-11e8-9b05-7f4d47baaea8.png)

2) running this consumer with a memory pool defined by
"buffer.memory"=30000000
I could limit the JVM to -Xmx120M -Xms120M

![image](https://user-images.githubusercontent.com/6661532/45823540-d6ad1480-bce5-11e8-9fd9-44051f6fe7e4.png)

edoardocomar commented:
As we expected, the new consumer setting `buffer.memory` appears more useful when 
- there is a need to constrain the overall memory usage of the consumer process
and
- the consumer is configured with a relatively large per-broker `fetch.max.bytes` 
and
- the consumer subscription is potentially spread among many brokers (or an unknown number of them)

mimaison commented:
@lindong28 @harshach @omkreddy @cmccabe Can one of you help with the review? Thanks

edoardocomar commented:
@hachikuji  we have rebased on trunk and would love feedback, thanks

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -640,7 +640,8 @@ public Node leastLoadedNode(long now) {
     public static AbstractResponse parseResponse(ByteBuffer responseBuffer, RequestHeader requestHeader) {
         Struct responseStruct = parseStructMaybeUpdateThrottleTimeMetrics(responseBuffer, requestHeader, null, 0);
-        return AbstractResponse.parseResponse(requestHeader.apiKey(), responseStruct);
+        // Sasl authentication does not use a MemoryPool, hence null Closeable
```
nit: it's not obvious that this method is used only for SASL. Perhaps we should consider either renaming it or moving it to a SASL-specific class?

mimaison replied:
Agreed. As `parseStructMaybeUpdateThrottleTimeMetrics` is not public, I've kept this method in `NetworkClient` but renamed it `parseSaslResponse`

hachikuji commented on clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java:
```diff
@@ -217,7 +217,8 @@ public Struct parseResponse(short version, ByteBuffer buffer) {
     public final Schema[] responseSchemas;
-    public final boolean requiresDelayedAllocation;
+    public final boolean requestRequiresDelayedAllocation;
+    public final boolean responseRequiresDelayedAllocation;
```
nit: I know this name was chosen for consistency, but in what sense does it require delayed allocation? It seems like really we're trying to control whether or not a response requires manual deallocation?

mimaison replied:
Some Response objects, like JoinGroupResponse, have fields directly backed by ByteBuffer. So even when the network buffer has been read into an object, we still need to keep it. This is what we're trying to keep track here. 
So deallocation of these can only be done once the Response object is not used anymore, instead of directly once the Response has been built (for Response without ByteBuffer fields)

hachikuji replied:
I understood the purpose of the flag, I was just nitpicking the name. Allocation happens immediately, right? I thought possibly `responseRequiresManualDeallocation` would be more accurate.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -464,6 +470,12 @@
+                                        Type.LONG,
+                                        Long.MAX_VALUE,
+                                        atLeast(1),
+                                        Importance.HIGH,
```
nit: I think the importance is probably low or medium. Most users don't have to think about this.

mimaison replied:
Agreed, we'll update that

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -131,7 +131,7 @@
             "this value, the record batch will still be returned to ensure that the consumer can make progress. As such, this is not a absolute maximum. " +
             "The maximum record batch size accepted by the broker is defined via <code>message.max.bytes</code> (broker config) or " +
-            "<code>max.message.bytes</code> (topic config). Note that the consumer performs multiple fetches in parallel.";
+            "<code>max.message.bytes</code> (topic config). Note that the consumer performs multiple fetches in parallel. It should be lower than <code>buffer.memory</code>.";
```
nit: `The value must be lower than <code>buffer.memory</code>.`

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -2220,4 +2234,39 @@ private void throwIfNoAssignorsConfigured() {
+        MetricName memoryPoolDepletedTimeMetricName = metrics.metricName("memory-pool-depleted-time-total", metricGrpName, "The amount of time when the MemoryPool is full");
+        memoryPoolSensor.add(new Meter(TimeUnit.MILLISECONDS, memoryPoolDepletedPercentMetricName, memoryPoolDepletedTimeMetricName));
+
+        int maxAllocation = (memoryPoolSize > Integer.MAX_VALUE) ? Integer.MAX_VALUE : (int) memoryPoolSize;
```
The default value is `Long.MAX_VALUE`, but this gets overridden to `Integer.MAX_VALUE` here. Is that intentional? If we really want a max cap at 2GB, maybe we should make the type of the config an int?

mimaison replied:
MemoryPool uses a long for `sizeInBytes` but an integer for the `maxAllocation`. By doing this we attempt at allowing single allocations to be as large as possible. So it's either the size of the map or `Integer.MAX_VALUE`

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -565,9 +571,13 @@ public void handle(JoinGroupResponse joinResponse, RequestFuture<ByteBuffer> fut
         try {
+            Map<String, ByteBuffer> clone = new HashMap<>(joinResponse.members().size());
+            for (Map.Entry<String, ByteBuffer> entry : joinResponse.members().entrySet()) {
+                clone.put(entry.getKey(), ByteBuffer.wrap(Utils.toArray(entry.getValue())));
```
Hmm.. This kind of defeats the point of the memory pool. Should we just not bother allocating these responses from the pool?

vahidhashemian replied:
It is possible to add a brief comment to explain this block of code? Also, one for the other clone a few lines below.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/Node.java:
```diff
@@ -28,20 +28,60 @@
+     * @param privileged If true, messages from this node will be allocated directly on the heap when the MemoryPool 
+     * is full
+     */
+    public Node(int id, String host, int port, boolean privileged) {
```
Hmm.. Unfortunately, `Node` is a public API. We probably need to find another way to propagate this information to the network layer.

mimaison replied:
This API change is covered by the KIP

hachikuji replied:
One new constructor is in the KIP, but do we still think it is a good idea? The notion of a privileged connection is only relevant to the internal implementation, so it would be unfortunate for it to leak into the public API.

On a related note, we were discussing whether it makes sense to allocate JoinGroup/SyncGroup using the buffer pool. I'm wondering if that suggests we should be able to control this allocation at the request level instead of the connection level. If we were doing it at the request level, then we could add an additional flag in `ClientRequest`. Then we wouldn't need the changes to `Node` here.

mimaison replied:
@hachikuji 

@edoardocomar and I have thought about your suggestion and we think we could get this to work without touching `Node`. The idea would be to rely on the request/response guaranteed ordering and match the incoming response with a pending request.

We could inject into `KafkaChannel`s an interface that would help it determine whether or not to read a response in the memoryPool - based on the apikey of the associated request.
The Consumer implementation of the interface would simply `peekLast()` the `inFlightRequests` collection held by `NetworkClient` and return whether the matching request requires allocation into the pool (i.e. whether it is a FETCH).
The helper interface could also be plugged into the `MemoryPool`.

That way, allocation would be determined by the apikey of the request instead of the node.

What do you think?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/AbstractResponse.java:
```diff
@@ -178,4 +184,14 @@ public int throttleTimeMs() {
         return toStruct(version).toString();
     }
+
+    public void close() {
```
nit: missing `@Override`

mimaison replied:
Good catch!

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/AbstractResponse.java:
```diff
@@ -68,12 +71,15 @@ protected void updateErrorCounts(Map<Errors, Integer> errorCounts, Errors error)
+    /*
+     * Responses that require a delayed allocation need a reference to their closeable NetworkReceive.
+     */
+    public static AbstractResponse parseResponse(ApiKeys apiKey, Struct struct, Closeable closeable) {
```
nit: I'm not sure we need or want the `Closeable` generality here. Could we just use the `NetworkReceive` type?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java:
```diff
@@ -541,4 +548,16 @@ private static void addPartitionData(String dest, Queue<Send> sends, Struct part
     }
+
+    /*
+     * The buffer underlying a FetchResponse is sliced into many CompletedFetch objects.
```
Kind of a meta comment, but this kind of bookkeeping seems like it would be better left inside `Fetcher`. This object is also used by the broker.

hachikuji reviewed (commented):
Apologies for the long delay. I left a few comments. See in particular the comments about JoinGroup/SyncGroup.

mimaison committed:
Addressed some comments

mimaison committed:
Addressed more comments

edoardocomar commented:
(bump) Hi @hachikuji 
please see
https://github.com/apache/kafka/pull/4934#discussion_r236339139
if the approach to use/skip the pool based on request key seems reasonable to you we're keen to come up with a new PR ...

mimaison commented:
@hachikuji  Gentle reminder =)

vahidhashemian commented on core/src/test/scala/integration/kafka/server/GssapiAuthenticationTest.scala:
```diff
@@ -35,8 +35,10 @@ import org.apache.kafka.common.security.auth.SecurityProtocol
+import org.apache.kafka.common.Node
 
 import scala.collection.JavaConverters._
+import org.apache.kafka.common.Node
```
nit: duplicate import

vahidhashemian commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -1208,6 +1198,12 @@ public void testOffsetOfPausedPartitions() {
         consumer.close();
     }
 
+    private Node prepareLookupCoordinator(MockClient client, Node node) {
```
nit: Perhaps `prepareCoordinatorLookup` is a better name?

vahidhashemian reviewed (commented):
Thanks for the KIP/PR. I left a few very minor comments.

mimaison commented:
@vahidhashemian Thanks for taking the time to review!

@hachikuji raised some interesting points in https://github.com/apache/kafka/pull/4934#discussion_r233266453. We've proposed a possible design in https://github.com/apache/kafka/pull/4934#discussion_r236339139. We've also started a very rough implementation in another branch: https://github.com/mimaison/kafka/tree/kip-81-wip.

@vahidhashemian @hachikuji Could you take a look at that branch and see if that method for tracking what to allocate in the pool would be viable? Keep in mind that's not a proper implementation but just a way to validate the method so we cut a number of corners to get it running as quickly as possible.

Thanks!

vahidhashemian commented:
@mimaison Thanks, yes I understand you and @hachikuji have gone through several rounds of review. Would it be possible to push the other branch as a separate PR (maybe link it to this one) and perhaps highlight the area with the new design? That should help with the review and finding the proposed changes. I'm assuming the majority of changes in the new PR is the same as this one. Thanks!

linking12 reviewed (changes requested):
some conflict， sorry for approve, Is 2.3 contain these?

mimaison commented:
@linking12 No this KIP is not in 2.3 and won't be in 2.4 either. 
This KIP is currently blocked as the implementation in this PR is not acceptable and the alternative implementation I propose https://github.com/apache/kafka/pull/6193 has not been reviewed.
